### PR TITLE
Funnels + Quadrotor Online Planning

### PR DESCRIPTION
**Pull request by psiorx**
This is the first pass at adapting Ani's funnel code to work with a different system and the new Drake.  

I've applied it to a quadrotor example that demonstrates online planning through a simulated forest using a trajectory library of my own design.  The trajectory library contains 5 locally optimal trajectories that are all sequentially composeable and maintain an average forward speed of 1 m/s throughout.  

The speed can be increased by simply dropping a different trajectory library in and/or adding new trajectories to the existing library.

Here's a visualization of the funnels built around the trajectories:

quad_funnels

...and a video showing the quadrotor applying these funnels online to negotiate the forest.

https://www.youtube.com/watch?v=mv4BH_5Sh_s&feature=youtu.be

### CODE REVIEW

**Comment by RussTedrake** [comment]
Awesome.  Will review tonight!

File: drake/addpath_drake.m [hunk 1]
```diff
@@ -40,6 +40,10 @@
 addpath(fullfile(root,'systems','frames'));
 addpath(fullfile(root,'systems','visualizers'));
 addpath(fullfile(root,'systems','robotInterfaces'));
+addpath(fullfile(root,'systems','onlinePlanning'));
```
Comment (RussTedrake) [inline drake/addpath_drake.m]: sorry it took so long for me to get to this.  we should definitely get it in. 

i don't love this name, or location.  are all of these tools really so distinct?  we already have a trajectory library class (which hasn't been hit much, and perhaps should be replaced by yours) in systems/trajectories.  the funnel libraries could go there, too.

File: drake/examples/Quadrotor/Quadrotor.m [hunk 2]
```diff
@@ -113,13 +113,13 @@ function visualizePlan(x,lcmgl)
   methods (Static)
     
     function runOpenLoop
-      r = Quadrotor('lidar');
+      r = Quadrotor();
```
Comment (RussTedrake) [inline drake/examples/Quadrotor/Quadrotor.m]: why this change?

File: drake/systems/onlinePlanning/FunnelLibraries/sampledFiniteTimeReach_B0.m [hunk 3]
```diff
@@ -0,0 +1,1506 @@
+function [V,rho,Phi] = sampledFiniteTimeReach_B0(sys,polyOrig,Vtraj0,G,R0,tv,ts,xtraj0,utraj,options,Phi,rho)
```
Comment (RussTedrake) [inline drake/systems/onlinePlanning/FunnelLibraries/sampledFiniteTimeReach_B0.m]: why not "sampledFiniteTimeReachability"

and I think this should definitely be a method of PolynomialSystem, like all of it's counterparts, not FunnelLibraries.

File: drake/systems/onlinePlanning/OnlinePlannerMex.m [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        t_start;
+        x_start;
+        init_replan;
+        
+    end
+    
+    methods
+        
+        % construction
+        function obj = OnlinePlannerMex(funnelLibrary,forestAhead,currentFunnel,cyclicIdx,t_replan,t_start,x_start,num_inputs)
```
Comment (RussTedrake) [inline drake/systems/onlinePlanning/OnlinePlannerMex.m]: forestAhead is pretty problem specific, when most other things in your class are pretty generic.

File: drake/systems/onlinePlanning/SharedDataHandleMutable.m [hunk 5]
```diff
@@ -0,0 +1,38 @@
+classdef SharedDataHandleMutable < handle
```
Comment (RussTedrake) [inline drake/systems/onlinePlanning/SharedDataHandleMutable.m]: why not just SharedDataHandle?

File: drake/systems/onlinePlanning/TrajectoryLibraries/TrajectoryLibrary.m [hunk 6]
```diff
@@ -0,0 +1,57 @@
+classdef TrajectoryLibrary
```
Comment (RussTedrake) [inline drake/systems/onlinePlanning/TrajectoryLibraries/TrajectoryLibrary.m]: i guess the old version was actually in dev: https://github.com/RobotLocomotion/drake/tree/master/drake/systems/trajectories/dev .  happy to have this replace that one.

File: drake/systems/onlinePlanning/Util/simulateEuler.m [hunk 7]
```diff
@@ -0,0 +1,31 @@
+function xtrajSim = simulateEuler(sysCl,tend,x0,dt,sys_type,xp,up)
```
Comment (RussTedrake) [inline drake/systems/onlinePlanning/Util/simulateEuler.m]: what?  no way.  you can call our existing simulate and set the solver to ode1.  don't duplicate code here...

Comment (psiorx) [inline drake/systems/onlinePlanning/Util/simulateEuler.m]: I've converted this over to use ode1 with the existing simulation pipeline but it needs the stuff in https://github.com/RobotLocomotion/drake/pull/1375

File: drake/examples/Quadrotor/Quadrotor.m (hunk 2 above)
Comment (psiorx) [inline drake/examples/Quadrotor/Quadrotor.m]: I turned this off to stop bullet from complaining.  With it on I was getting this:
```
In BulletModel::findClosestPointsBtwElements: No closest point found!
```

File: drake/systems/onlinePlanning/OnlinePlannerMex.m (hunk 4 above)
Comment (psiorx) [inline drake/systems/onlinePlanning/OnlinePlannerMex.m]: Agreed.

File: drake/examples/Quadrotor/Quadrotor.m (hunk 2 above)
Comment (RussTedrake) [inline drake/examples/Quadrotor/Quadrotor.m]: as reported in https://github.com/RobotLocomotion/drake/issues/1378 <https://github.com/RobotLocomotion/drake/issues/1378>

File: drake/systems/onlinePlanning/SharedDataHandleMutable.m (hunk 5 above)
Comment (psiorx) [inline drake/systems/onlinePlanning/SharedDataHandleMutable.m]: Some of the wrapped data gets modified in mex but it looks like SharedDataHandle still works with that case.  I'll zap the Mutable variant.

File: drake/examples/Quadrotor/Quadrotor.m (hunk 2 above)
Comment (RussTedrake) [inline drake/examples/Quadrotor/Quadrotor.m]: but i really want to leave the lidar on so it’s getting coverage on the build servers…

Comment (psiorx) [inline drake/examples/Quadrotor/Quadrotor.m]: > i really want to leave the lidar on so it’s getting coverage

I agree.  

I think I made this change early on while hacking around in the quadrotor examples.  In any case, this PR doesn't even call Quadrotor.runOpenLoop so it shouldn't be here.

**Comment by psiorx** [comment]
Is this ready to go?

File: drake/systems/trajectories/isCollisionFree_mex.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+
+/* Shift and transform vertices
+ */
+double *shiftAndTransform(double *verts, double *vertsT, const mxArray *x, mxArray *x0, int k, mxArray *cSk, size_t nRows, size_t nCols)
+{
+    /* This can and maybe should be sped up. For example, we know that cSk is upper triangular.*/
+    double *dcSk = mxGetPr(cSk);
+    double *dx0 = mxGetPr(x0);
+    double *dx = mxGetPr(x);
```
Comment (RussTedrake) [inline drake/systems/trajectories/isCollisionFree_mex.cpp]: These should really be mxGetPrSafe
Nonenof the mex code in this request really uses our safety wrappers.

File: drake/systems/trajectories/ptToPolyBullet_mex.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+We could probably make the radius 0 and be ok, but I'm not sure if bullet expects things to be non-degenrate.
+*/
+static const double radius = 0.5;
+static btSphereShape* point = new btSphereShape(radius); 
+
+void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
+{
+	 // Check for proper number of arguments.
+ 	 if(nrhs!=1) {
+	   mexErrMsgIdAndTxt( "MATLAB:ptToPolyDist:invalidNumInputs",
```
Comment (RussTedrake) [inline drake/systems/trajectories/ptToPolyBullet_mex.cpp]: Not "MATLAB:"

File: drake/systems/trajectories/replanFunnels_mex.cpp [hunk 10]
```diff
@@ -0,0 +1,1090 @@
...
+    
+    // lenrw, leniw, lencw
+    const int DEFAULT_LENRW = 500000;
+    const int DEFAULT_LENIW = 500000;
+    const int DEFAULT_LENCW = 500;
+    
+    snopt::integer lenrw = DEFAULT_LENRW, leniw = DEFAULT_LENIW, lencw = DEFAULT_LENCW;
+    
+    // cw, iw, rw
+    snopt::doublereal rw_static[DEFAULT_LENRW];
```
Comment (RussTedrake) [inline drake/systems/trajectories/replanFunnels_mex.cpp]: this should use the updated snopt heap allocation version, as seen in NonlinearProgramSnoptcmex

File: drake/systems/trajectories/shiftFunnel_snopt_mex.cpp [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+    
+    // lenrw, leniw, lencw
+    const int DEFAULT_LENRW = 500000;
+    const int DEFAULT_LENIW = 500000;
+    const int DEFAULT_LENCW = 500;
+    
+    snopt::integer lenrw = DEFAULT_LENRW, leniw = DEFAULT_LENIW, lencw = DEFAULT_LENCW;
+    
+    // cw, iw, rw
+    snopt::doublereal rw_static[DEFAULT_LENRW];
```
Comment (RussTedrake) [inline drake/systems/trajectories/shiftFunnel_snopt_mex.cpp]: again.  allocate to the heap instead of the stack

**Comment by RussTedrake** [comment]
you could also add the image/video to the gallery.
